Auto crop

sxzhang25 opened the pull request:
Implement a simple crop-transform algorithm that approximates frames with bounding boxes.

sxzhang25 committed:
Add fixed crop scale.

sxzhang25 committed:
Use bounding box intersection approach to find best crop window.

sxzhang25 committed:
Add additional zoom to account for bounding box approximation errors.

sxzhang25 committed:
Merge branch 'stabilization-v2' into auto-crop

# Conflicts:
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java

sxzhang25 committed:
Merge branch 'refactor-reader-utils' into auto-crop

sxzhang25 committed:
Add option in widget to enable or disable auto-crop.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,82 @@
+
+    private static final String TAG = "BoundingBox";
+
+    public float xMin;
+    public float yMin;
+    public float xMax;
+    public float yMax;
```
It seems that you could make these final?

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,82 @@
+
+    private float error;
+
+    public BoundingBox() {
+        xMin = -1.0f;
+        yMin = -1.0f;
+        xMax = 1.0f;
+        yMax = 1.0f;
+        error = 0.0f;
+    }
```
This constructor can be deleted.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,82 @@
+        error = 0.0f;
+    }
+
+    public BoundingBox(Float[] a, Float[] b, Float[] c, Float[] d) {
```
Please add Javadoc for arguments a, b, c, d.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,82 @@
+        return yMax - yMin;
+    }
+
+    public float error() {
```
What is `error`? Please add Javadoc. (same for other methods)

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java:
```diff
@@ -37,6 +40,14 @@
+    /**
+     * Coordinates for corners of frame in OpenGL position coordinate system.
+     */
+    public static final Float[] BOTTOM_LEFT  = {-1.0f, -1.0f, 1.0f};
```
Is there any reason why you are using Float[] instead of float[] everywhere? Boxed types are way more expensive than their unboxed counterparts.

sxzhang25 replied:
The stabilization data in the video track is passed as a List<Float> to represent each homography, so I just kept using Float for consistency. I've changed everything over to float[] though, since nothing actually requires them to be boxed types.

sxzhang25 committed:
Minor fixes:

- Change all Float[] arrays to float[] arrays
- Add Javadoc to BoundingBox.java
- Create new constructor in BoundingBox.java to directly construct a bounding box (with the quadrilateral assumed to be the same as the bounding box)

clju2 commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,124 @@
+
+    /**
+     * Constructor for a bounding box using a given quadrilateral.
+     * The arguments a, b, c, d are all homogeneous 2D coordinates (x, y, 1) of the vertices of the
```
If the third coordinate is always 1, why pass it? Furthermore, it seems that the coordinates are not treated as homogenous inside the code (that is, you are not dividing the first and second coordinate by the third).

sxzhang25 replied:
Most of the coordinates I use in the rest of the API are homogeneous coordinates with z=1 (since they have to be multiplied by a bunch of 3x3 homography matrices) so I thought it would be easiest to just have a constructor that specified the coordinates are homogeneous. The homography matrices also never change the value of z so I never divided the z-coordinate out anywhere. There's nothing in the BoundingBox code that needs the extra z-dimension though, so should I just remove this Javadoc line?

Also, I could pass in 2-dimensional coordinates but I would have to create another 3-dimensional coordinate for each point when I multiply them by homography matrices again, so I figured I should just use the 3-dimensional coordinates everywhere. Should I still just switch over to non-homogeneous coordinates in this case?

clju2 replied:
I see. If we had time I would probably still recommend to change it (because you are essentially passing useless information that might even cause errors in the future if people start treating it like actual homogeneous coordinates).

One example of problems that might arise is that you are manipulating these homogeneous coordinates with 3x3 homographies, which might of course change the third coordinate if perspective-like transforms are applied. These resulting homogeneous coordinates will not have w = 1 anymore, and therefore will cause issues with functions such as the one we are commenting on.

One solution could be to have 2D (x, y) coordinates everywhere, and refactor your homography multiplication code to take a 2D vector and return a 2D vector (if need be, divide by the third coordinate as a last step in that function).

But for now, as I said, since we're out of time, we can keep it like that.

Another solution is to treat them like real homogeneous coordinates everywhere.

sxzhang25 replied:
Ok I see--I'll make a note of this in my final documentation. Thanks!

clju2 commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,124 @@
+                Math.min(this.xMax, other.xMax),
+                Math.min(this.yMax, other.yMax)
+        );
+        return intersection;
```
nit: no need to have temp variable `intersection`.

clju2 commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,124 @@
+    public final float xMax;
+    public final float yMax;
+
+    /**
+     * A pseudo-measurement of how well the bounding box approximates the quadrilateral. This is
+     * computed by taking the maximum distance from a vertex of the quadrilateral to a neighboring
+     * vertex of the bounding box. Thus, the error lies in the range [0, max(W, H) / 2], where
+     * W is the width of the bounding box and H is the height of the bounding box.
+     */
```
Could you put the explanation of this field in the Javadoc of the `error` function? This comment on a private field is not visible by users of this class.

clju2 commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -296,6 +270,38 @@ private int compileShader(int type, String shaderSrc) {
+                /* offset = */ 0
+        );
+        if (glGetError() != 0) {
+            throw new RuntimeException("Failed to get matrix");
```
Maybe `Log.e` instead? In the worst case it will result in visual artifacts, maybe we should not crash over it?

clju2 replied:
It seems that you still have a lot of these RuntimeExceptions thrown when GL returns errors, is that deliberate?

sxzhang25 replied:
No, these should have been changed. I didn't see these comments the first time around--sorry about that!

clju2 replied:
No problem!

clju2 commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java:
```diff
@@ -377,7 +383,7 @@ private void storeAsGLMatrix(HomographyMatrix homography) {
     public void drawFrame(List<HomographyMatrix> homographyList) {
         Log.d(TAG, "Drawing frame");
-        Log.d(TAG, "Homography list size: " + homographyList.size());
+
```
Do you want to leave the `Log.d` message above? Seems like it will spam logcat.

clju2 replied:
I think this was not addressed?

sxzhang25 replied:
Went through and deleted all the unnecessary Log.d(...) messages. All the files should be cleaned up now in terms of logging.

clju2 commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java:
```diff
@@ -0,0 +1,124 @@
+        xMax = Math.max(Math.max(a[0], b[0]), Math.max(c[0], d[0]));
+        yMax = Math.max(Math.max(a[1], b[1]), Math.max(c[1], d[1]));
+
+        float errorA = Math.min(Math.abs(a[0] - xMin), Math.abs(a[0] - xMax));
+        if (errorA < EPS) {
+            errorA = Math.min(Math.abs(a[1] - yMin), Math.abs(a[1] - yMax));
+        }
+
+        float errorB = Math.min(Math.abs(b[0] - xMin), Math.abs(b[0] - xMax));
+        if (errorB < EPS) {
+            errorB = Math.min(Math.abs(b[1] - yMin), Math.abs(b[1] - yMax));
+        }
+
+        float errorC = Math.min(Math.abs(c[0] - xMin), Math.abs(c[0] - xMax));
+        if (errorC < EPS) {
+            errorC = Math.min(Math.abs(c[1] - yMin), Math.abs(c[1] - yMax));
+        }
+
+        float errorD = Math.min(Math.abs(d[0] - xMin), Math.abs(d[0] - xMax));
+        if (errorD < EPS) {
+            errorD = Math.min(Math.abs(d[1] - yMin), Math.abs(d[1] - yMax));
+        }
+
+        error = (float) Math.sqrt(2.0) *
+                Math.max(Math.max(errorA, errorB), Math.max(errorC, errorD));
```
What will you use this error metrics for? It seems that it can still be very small (around `EPS`) for some quadrilateral that are not at all well approximated by their bounding box.

sxzhang25 replied:
So part of auto-crop was to re-center and re-scale the entire stabilized video so that the edges of the frames don't show in the viewport (without auto-crop, the video is stabilized but you can see the frame moving around in the viewport). I used BoundingBoxes to approximate the frames so that it would be easier to calculate how much I would need to translate and scale the whole video so that I could avoid seeing the edges of the frame as much as possible (and have the video fill up the whole view). But for a frame that is super rotated (like a diamond), the bounding box is a pretty bad approximation. So I calculated an error that (very loosely) approximated the longest distance from any vertex of the bounding box to an edge of the actual frame, and I use the error to increase the scale by some factor, just to hedge the possibility of bad approximations a little bit. It didn't work perfectly for all videos, but it did actually help in some super shaky videos.

It makes the auto-crop process a little more complicated though. Do you think it's worth keeping, or should I just stick with a simple approach?

clju2 replied:
Up to you! I guess you can keep this heuristic if if works in practice, but I just wanted to mention that I definitely found cases where your metrics would return a very small value (indicating that the fit is good) even though it's really not. Once again, since we're out of time, feel free to not do anything here, but maybe add in the Javadoc that it's a approximate heuristic, not valid in all cases.

sxzhang25 replied:
Ok I think I see the cases that you are talking about (e.g. thin diagonal rectangle with all vertices close to two opposite vertices of the bounding box). I think I'll get rid of the error measure then. Maybe something like using the difference in area between the bounding box and the quadrilateral would have been better? But it seems like a good move to remove it for now since it wasn't adding too much benefit.

clju2 replied:
Difference in area sounds like a good metrics to me :)

sxzhang25 committed:
Merge branch 'master' into auto-crop

Update branch to include code review changes.

sxzhang25 committed:
Fix code formatting.

sxzhang25 committed:
Clean up code and remove unneeded features.

- Clean up unnecessary logcat statements.
- Removed error field in BoundingBox, as it was providing limited added benefit.
- Removed distanceTo() function in HomographyMatrix (not a function that deals with homography matrices).
- Optimized imports.
- Fixed MotionPhotoUtilsTest to reflect MotionPhotoReader.open(...) parameter changes.
